bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/storage/directentrylogger/DirectWriter.java

```diff
@@ -0,0 +1,322 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.bookkeeper.bookie.storage.directentrylogger;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static org.apache.bookkeeper.common.util.ExceptionMessageHelper.exMsg;
+
+import io.netty.buffer.ByteBuf;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import org.apache.bookkeeper.common.util.nativeio.NativeIO;
+import org.apache.bookkeeper.common.util.nativeio.NativeIOException;
+import org.apache.bookkeeper.slogger.Slogger;
+import org.apache.commons.lang3.SystemUtils;
+
+class DirectWriter implements LogWriter {
+    final NativeIO nativeIO;
+    final int fd;
+    final int id;
+    final String filename;
+    final BufferPool bufferPool;
+    final ExecutorService writeExecutor;
+    final Object bufferLock = new Object();
+    final List<Future<?>> outstandingWrites = new ArrayList<Future<?>>();
+    Buffer nativeBuffer;
+    long offset;
+    private volatile boolean useFallocate = true;
+
+    DirectWriter(int id,
+                 String filename,
+                 long maxFileSize,
+                 ExecutorService writeExecutor,
+                 BufferPool bufferPool,
+                 NativeIO nativeIO, Slogger slog) throws IOException {
+        checkArgument(maxFileSize > 0, "Max file size (%d) must be positive");
+        this.id = id;
+        this.filename = filename;
+        this.writeExecutor = writeExecutor;
+        this.nativeIO = nativeIO;
+
+        offset = 0;
+
+        try {
+            fd = nativeIO.open(filename,
+                               NativeIO.O_CREAT | NativeIO.O_WRONLY | NativeIO.O_DIRECT,
+                               00755);
```

Comment:
`00755` as in permissions? why 755 and not e.g. 600 or 640?
"755 means read and execute access for everyone and also write access for the owner of the file." 
Why are we granting execute access to this file? 
Why read access is needed to everyone?

Reply:
`644` is better. The current ledger log file's permission is `644`, we can keep the same as the previous one.